esmvalcore/esgf/_search.py

```diff
@@ -52,14 +50,24 @@ def same_file(file):
         dataset = file.dataset.rsplit('.', 1)[0]
         return (dataset, file.name)
 
+    if isinstance(versions, str):
+        versions = (versions, )
+
     files = sorted(files, key=same_file)
-    for _, versions in itertools.groupby(files, key=same_file):
-        versions = sorted(versions, reverse=True)
-        latest_version = versions[0]
+    for _, group in itertools.groupby(files, key=same_file):
+        group = sorted(group, reverse=True)
+        if versions:
+            selection = [f for f in group if f.facets['version'] in versions]
+            if not selection:
+                raise FileNotFoundError(
+                    f"Requested versions {', '.join(versions)} of file not "
+                    f"found. Available files: {group}")
+            group = selection
+        latest_version = group[0]
         result.append(latest_version)
-        if len(versions) > 1:
+        if len(group) > 1:
             logger.debug("Only using the latest version %s, not %s",
-                         latest_version, versions[1:])
+                         latest_version, group[1:])
```

Comment:
I was thinking if we want an option that stops the download if the version requirements are not met; something in the `facets` that is set to False by default, but the user can set it to True and hence don't download the latest version if files with their required versions are not found. Also, how is this case handled: say a user wants version vxx that is not available temporarily (ESGF node where that data is is briefly offline), does this mean they will get `latest` instead with no questions asked?

Reply:
This function is used to search ESGF index nodes for files, it does not download files. Note that an index node is either online and then you can use it to search for files, or it's offline and then you cannot use it at all and the code will automatically try to use another index node. I'm not sure how well that works in practice though because search is somehow distributed over index nodes ([using distrib=true](https://esgf.github.io/esg-search/ESGF_Search_RESTful_API.html)), so I think that if one index node is down you might indeed miss results without noticing.

>  Also, how is this case handled: say a user wants version vxx that is not available temporarily (..), does this mean they will get latest instead with no questions asked?

If a user requests version `'vxx'` and it's not available in the search index, a `FileNotFoundError` will be raised. Only if the user does not specify which version(s) they want will they get the latest version(s).

Reply:
aha, the correct behaviour I was after, thanks, man! :beer: